### DISCUSSION TITLE
Initialize an atom with subscriptions

### DISCUSSION THREAD

**Discussion by user72356** [pmndrs/jotai#1702] [Q&A] [answered]
I'm using Supabase (a DB service + an auth provider). A Supabase client must be created in order to authenticate users. Essentially I instantiate the client and immediately after that I set up subscriptions that update an `authSession` variable whenever an auth event occurs.

At the moment, I'm able to properly do this in Zustand. My `supabase` store variable gets initialized using the following function, and the `authSession` store variable is set whenever an auth event occurs.

How can I do this in Jotai? I can create the client in the initializer of a "supabase" atom, but I can't set the subs since I would need write access to an "authSession" atom, which I don't because I'm in an initializer. I suppose I could set up the subs in a write atom (action atom), but this action would have to be called from somewhere. This need to be called as close to the start of the app lifecycle as possible.

Thank you!

```typescript
function createSupabaseClient(persistSession: boolean): SupabaseClient
{
    console.log("createSupabaseClient");

    const supabase = createClient(
        import.meta.env.VITE_SUPABASE_URL,
        import.meta.env.VITE_SUPABASE_ANON_KEY,
        { auth: { persistSession } }
    );

    supabase.auth.onAuthStateChange((event, session) =>
    {
        if ((event === "SIGNED_IN" || event === "TOKEN_REFRESHED") && session !== null)
        {
            set({ authSession: session });
        }
        else if (event === "SIGNED_OUT")
        {
            set({ authSession: undefined });
        }
        else if (event === "PASSWORD_RECOVERY")
        {
            set({
                isAuthDialogOpen: true,
                currentAuthDialogView: AuthDialogView.changepassword
            });
        }
    });

    return supabase;
}
```

**Comment by dai-shi** [comment]
Jotai is modeled after useState+useContext use case. So, in your mental model, you want to think how it will be done with useState. From there, using useAtom should be straightforward.

(Jotai v2 API will provide another feature so that it can be used like zustand, but we are not sure yet how the best practice will be like.)

**Reply by user72356** [reply]
What is the v2 feature? Is this something I can experiment with today?

**Reply by dai-shi** [reply]
It's pre-released since `jotai@1.11.0`.
https://jotai.org/docs/guides/migrating-to-v2-api

**Reply by user72356** [reply]
> Jotai v2 API will provide another feature so that it can be used like zustand

Were you referring to the [Store API](https://jotai.org/docs/guides/migrating-to-v2-api#store-api)?

**Reply by user72356** [reply]
Found the fix using the Store API as you alluded to. I will post that.

**Comment by user72356** [comment, marked as answer]
Turns out Jotai v2 can handle this case quite nicely as @dai-shi alluded to. 

I ended up creating a supabase atom and then an action atom (write-only atom) that initializes the supabase atom. It's akin to an action reducer function in Zustand or Redux. This you can do with v1.

Where v2 comes in, is that I can create and manipulate the store outside of React, meaning I can call that initialization action at the module level. This way Supabase is initialized (including the subscriptions) as early as possible in the app lifecycle. And it works great.

At the module level, just use `createStore()` and then `store.set()` to call your init action atom. Don't forget to wrap your React tree with a `<Provider>` to pass that manually created store down into your components. The beauty is that everything is transparent at the component level and you don't need to refer to that manual store at all. It just works.
